HPCC-19553 Generate repeatable attributes as a list format

shamser opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

shamser committed:
HPCC-19553 Generate repeatable attributes as a list format

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8ab9eda9cb3ac2ff764b77e71a7e61a1a82421f7
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |784 |784 |0 |
|test (thor) |715 |715 |0 |
|test (roxie) |828 |828 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  168 sec (00:02:48) |   61 sec (00:01:01) |    8 sec (00:00:08) |   29 sec (00:00:29) | 1228 sec (00:20:28) |   20 sec (00:00:20) | 1534 sec (00:25:34) |

shamser commented:
@ghalliday

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: ac04e3d543e30312f148b39ead751abc89c078d3
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |784 |784 |0 |
|test (thor) |715 |715 |0 |
|test (roxie) |828 |828 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  174 sec (00:02:54) |   60 sec (00:01:00) |    9 sec (00:00:09) |   29 sec (00:00:29) | 1199 sec (00:19:59) |   19 sec (00:00:19) | 1508 sec (00:25:08) |

ghalliday commented on testing/esp/wudetails/key/dict1_hthor_testcase1.json:
```diff
@@ -459,23 +459,23 @@ testCase(ScopeFilter={
                         },
                         {
-                            'Name': 'Definition',
+                            'Name': 'DefinitionList',
```
Note - the repeated attributes may not occur next to each other.  In this case this definition should have been combined with the attribute on line 462.

ghalliday reviewed (commented):
@shamser I think you might need to rethink the implementation a bit because the attributes can come out in any order (and do from a graph).

ghalliday commented:
The other alternative is to ensure they are reported together, but I'm not quite sure who you would do that.

shamser commented:
@ghalliday Fixed: the lists are combined even if the attributes come in random order.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5ea8a49e16e4e9a39c665ba9750f344a1a4bd2bb
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |786 |786 |0 |
|test (thor) |717 |717 |0 |
|test (roxie) |830 |830 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  169 sec (00:02:49) |   61 sec (00:01:01) |    6 sec (00:00:06) |   27 sec (00:00:27) | 1185 sec (00:19:45) |   19 sec (00:00:19) | 1482 sec (00:24:42) |

ghalliday commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -64,10 +64,79 @@ class WUDetailsVisitor : public IWuScopeVisitor
+        {
+            AttribValuePair tmp(attr,value);
+            attribValuePairs.insert(tmp);
+        };
```
minor: extra ;

ghalliday commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -64,10 +64,79 @@ class WUDetailsVisitor : public IWuScopeVisitor
+            attribValuePairs.clear();
+        }
+    private:
+        struct AttribValuePair
```
could use

    typedef std::pair<WuAttr, StringBuffer> AttribValuePair;

from \<utility>

ghalliday commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -122,32 +192,49 @@ void WUDetailsVisitor::noteStatistic(StatisticKind kind, unsigned __int64 value,
+    bool returnList = (listAttr==WaNone)?false:true;
+    bool returnAttr = true;
+
+    if (extraAttributesRequested || returnAttributeSpecified)
```
I don't expect it is wrong, but I can't follow the logic of this code.  That suggests it should be split up/ refactored into some functions.  E.g.,

```
if (include(attr))
   addAttrib
else if (listAttr != None && include(listAttr))
   addMultiListItem();

```

ghalliday reviewed (commented):
@shamser a few minor comments.  Nothing obviously wrong, but I couldn't follow some of the code which suggests it could be cleaned up.

ghalliday commented:
@shamser

shamser commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -64,10 +64,79 @@ class WUDetailsVisitor : public IWuScopeVisitor
+            WuAttr attr;
+            StringBuffer values;
+
+            AttribValuePair(WuAttr _attr, const char * value): attr(_attr), values(value) {};
```
A mistake now but as a result of the first version maintaining a coma delimited values, rather than single values.

shamser commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -64,10 +64,79 @@ class WUDetailsVisitor : public IWuScopeVisitor
+    class MultiListValues
+    {
+    public:
+        MultiListValues(WUDetailsVisitor & _parent):parent(_parent){};
```
Changed.

shamser commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -209,6 +296,26 @@ void WUDetailsVisitor::buildAttribListToReturn(IConstWUPropertiesToReturn & prop
 
+    bool attributeListSeparateRequested = false;
+    StringArray & propertiesToReturnList = propertiesToReturn.getProperties();
+    ForEachItemIn(i,propertiesToReturnList)
```
I've refactored, simplified and documented the code to make it easier to follow.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 0d95cc7094e422ccb6eae7d5155d9a4371b04a9e
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |786 |786 |0 |
|test (thor) |717 |717 |0 |
|test (roxie) |830 |830 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  177 sec (00:02:57) |   61 sec (00:01:01) |    8 sec (00:00:08) |   29 sec (00:00:29) | 1189 sec (00:19:49) |   19 sec (00:00:19) | 1493 sec (00:24:53) |

shamser commented:
@ghalliday

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 56c334fbae91075a655314aeded33e140d5c1b15
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |786 |786 |0 |
|test (thor) |717 |717 |0 |
|test (roxie) |830 |830 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  173 sec (00:02:53) |   61 sec (00:01:01) |    9 sec (00:00:09) |   29 sec (00:00:29) | 1168 sec (00:19:28) |   19 sec (00:00:19) | 1477 sec (00:24:37) |

ghalliday reviewed (commented):
A few comments

ghalliday commented:
@shamser the new code was definitely easier to follow.  A few comments

shamser commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -120,39 +182,74 @@ void WUDetailsVisitor::noteStatistic(StatisticKind kind, unsigned __int64 value,
+{
+    if (propertiesToReturnAttribs->test(attr-WaNone))
+        return true;
+    if (extraAttributesForCurScope==nullptr &&
```
extraAttributeForCurScope points to the list of attributes to be returned for the_current_ scope type - it's updated every time the scope changes.   There may be scope types for which extra attributes is not necessary. In which case, extraAttributesForCurScope will be nullptr.

ghalliday replied:
The code still doesn't look valid.
if extraAttributesForCurScope==nullptr then it is checking that

    nullptr->find(attr)==nullptr->end()

which is not what you want.

shamser commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -161,14 +258,29 @@ void WUDetailsVisitor::noteHint(const char * kind, const char * value)
+// When neither possible, return false.
+static bool getPropertyIdFromName(const char *propName, StatisticKind & sk, WuAttr & wa )
+{
+    sk = queryStatisticKind(propName, StKindNone);
```
The original code is actually correct: it only attempts to initialize wa if it can't find a valid name as StatisticKind.

However, I have refactored to make the code clearer.

ghalliday replied:
I agree the code is correct for the current usage, but if you happened to call the same function in a different content you could get confusing results.

ghalliday replied:
You should either cleanly initialise wa (preferred) or document in the function description that you need to check sk before wa.

shamser commented:
@ghalliday I have updated the code. Please note that I didn't need to change 2 areas as the original seems correct  (I've replied to those 2 review comments.)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 1c4237892679138695f5952558317dd32c9cb7b9
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  170 sec (00:02:50) |   61 sec (00:01:01) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1208 sec (00:20:08) |   18 sec (00:00:18) | 1513 sec (00:25:13) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5ef52b846ef5cd82b78ae10d79f6c26e55c8af35
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  175 sec (00:02:55) |   60 sec (00:01:00) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1212 sec (00:20:12) |   19 sec (00:00:19) | 1523 sec (00:25:23) |

ghalliday commented:
@shamser two replies to your comments/changes.

shamser commented:
@ghalliday I've addressed your comments and found couple more issues that I had to fix:
1) Multiset was causing multi list items were being returned in random order; even different order from run to run.  This caused problem with the regression script.
2) There was a bug in the code: the multi kind of an attribute was always returned even if the single kind was requested. e.g even if IdDependency was requested in PropertiesToReturn, IdDependencyList was returned.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 7e1969df75a626989a58b48dcfa10bc8bb46890f
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  172 sec (00:02:52) |   61 sec (00:01:01) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1381 sec (00:23:01) |   18 sec (00:00:18) | 1693 sec (00:28:13) |

ghalliday commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -70,36 +76,39 @@ class WUDetailsVisitor : public IWuScopeVisitor
+            {
+                StringBuffer newval ("[");
+                newval.appendf("\"%s\"",  value);
+                AttribValuePair newvalpair(attr, newval);
```
For a little bit more efficiency you can use emplace_back avoiding the local object.

ghalliday replied:
You could alternatively append the value to temp.second and insert that.

ghalliday commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -327,6 +338,25 @@ void WUDetailsVisitor::buildAttribListToReturn(IConstWUPropertiesToReturn & prop
+    // Examine propertiesToReturn to see if any thing there will require a post filter
+    if (!attributesPostFilterRequired)
+    {
+        // If singleKind is specifically requested, it must be handled by post filter
```
Slightly confusing, and worth clarifying:  It doesn't require a post filter, but the post filter code also controls whether list items are generated as single.

ghalliday reviewed (commented):
A few more comments!

ghalliday commented:
@shamser

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 16c420a62c10c5208951249399eb67cb485d68ee
Build: success
ECL Watch:  Rebuilding Site
Rebuild: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |829 |2 |


[Errors:](https://gist.github.com/f70e8b16b459d95c1f8cca1282bc1d9f)
- test
  - roxie
    - 559\. [roxiepipe](https://gist.github.com/HPCCSmoketest/f70e8b16b459d95c1f8cca1282bc1d9f#file-roxiepipe-w20180613-105040-1-log), 604\. [soapcall (targetIP=127001, goodPort=9876, blacListedPort=9875)](https://gist.github.com/HPCCSmoketest/f70e8b16b459d95c1f8cca1282bc1d9f#file-soapcall-targetip-127001,-goodport-9876,-blaclistedport-9875-w20180613-105057-1-log), 

HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  175 sec (00:02:55) |  125 sec (00:02:05) |   10 sec (00:00:10) |   31 sec (00:00:31) | 1351 sec (00:22:31) |   19 sec (00:00:19) | 1733 sec (00:28:53) |

shamser commented:
@ghalliday

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8e10ea4d824d3a07118b43068eeca9eb7df4c19c
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  171 sec (00:02:51) |   61 sec (00:01:01) |    9 sec (00:00:09) |   29 sec (00:00:29) | 1168 sec (00:19:28) |   19 sec (00:00:19) | 1474 sec (00:24:34) |

ghalliday commented on esp/services/ws_workunits/ws_wudetails.cpp:
```diff
@@ -48,22 +54,73 @@ class WUDetailsVisitor : public IWuScopeVisitor
+
+        void flushListAttributes(WUDetailsVisitor & parent)
+        {
+            StringBuffer values;
```
trivial: unused local

ghalliday commented:
@shamser please delete that one unused variable (coverity will complain) and then squash and I will merge

shamser commented:
@ghalliday Unused variable removed and squashed.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 49832c6592c01788871f109cc614f4dea104b03b
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   11 sec (00:00:11) |  171 sec (00:02:51) |   61 sec (00:01:01) |    8 sec (00:00:08) |   29 sec (00:00:29) | 1378 sec (00:22:58) |   18 sec (00:00:18) | 1676 sec (00:27:56) |
